### ISSUE TITLE
Env Vars not passed to build init container

### ISSUE THREAD

**Issue by georgethebeatle** [pivotal/kpack#995]
We are attempting to deploy [postfacto](https://github.com/pivotal/postfacto) using kpack deplyed in [korifi](https://github.com/cloudfoundry/korifi).

We get a failure in the build init container in the Rails Assets buildpack:
```
Paketo Rails Assets Buildpack 0.6.0
  Executing build process
    Running 'bundle exec rails assets:precompile assets:clean'
      rails aborted!
      Bundler::GemRequireError: There was an error while trying to load the gem 'friendly_id'.
      Gem Load Error is: Database URL cannot be empty
```

It is arguable that DATABASE_URL should not be required to build templates, but that is how the app is.

If we run the build process with `pack`, we see a similar but different error in the Rails Assets buildpack if we don't pass any env vars:

```
Paketo Rails Assets Buildpack 0.6.0
  Executing build process
    Running 'bundle exec rails assets:precompile assets:clean'
      rails aborted!
      ActiveRecord::AdapterNotSpecified: The `production` database is not configured for the `production` environment.

        Available databases configurations are:

        development
      test
      /workspace/config/environment.rb:35:in `<top (required)>'
      Tasks: TOP => environment
      (See full trace by running task with --trace)
```

However, if we pass DATABASE_URL through `--env` in pack, it allows the Rails Assets buildpack to succeed and the whole build succeeds.

So we believe it is intentional that env vars supplied to the build are accessible in the build step, but kpack doesn't appear to be doing that.

To summarise, we think that env vars set in spec.env of the kpack build resource should be visible to the buildpacks in the kpack build container. Could you confirm whether this is correct, or if there is another way we should be making this happen?

Thanks,

Georgi and @kieron-dev

**Comment by tylerphelan** [comment]
@georgethebeatle 

Env vars should be provided to the `build` container.

>We get a failure in the build init container
The log provided suggests that this error is in the `build` container.

> So we believe it is intentional that env vars supplied to the build are accessible in the build step, but kpack doesn't appear to be doing that.

To validate this:
1. check that the `image` resource `spec.build.env` is set
2. Could you check the `env` of the build pod `build` container is unset

**Comment by georgethebeatle** [comment]
Hi @tylerphelan

Here is an example image from our cluster:

```
apiVersion: kpack.io/v1alpha2
kind: Image
metadata:
  creationTimestamp: "2022-07-19T14:42:31Z"
  generation: 1
  labels:
    korifi.cloudfoundry.org/build-workload-name: 42c5aa03-9d5e-416b-ae92-3aaba8603547
  name: 42c5aa03-9d5e-416b-ae92-3aaba8603547
  namespace: cf-space-4da0295e-53a9-4500-a15e-576aee052999
  ownerReferences:
  - apiVersion: korifi.cloudfoundry.org/v1alpha1
    kind: BuildWorkload
    name: 42c5aa03-9d5e-416b-ae92-3aaba8603547
    uid: 2f52f5d3-fccb-407f-9389-78bade9a472b
  resourceVersion: "10254081"
  uid: 5eecabc7-0fb5-48b7-84c9-4e304caf9be1
spec:
  build:
    env:
    - name: ACTION_CABLE_HOST
      valueFrom:
        secretKeyRef:
          key: ACTION_CABLE_HOST
          name: 0403724e-d0b5-4edb-ad16-dc310d22b771-env
    - name: DATABASE_URL
      valueFrom:
        secretKeyRef:
          key: DATABASE_URL
          name: 0403724e-d0b5-4edb-ad16-dc310d22b771-env
    - name: SECRET_KEY_BASE
      valueFrom:
        secretKeyRef:
          key: SECRET_KEY_BASE
          name: 0403724e-d0b5-4edb-ad16-dc310d22b771-env
    - name: SESSION_TIME
      valueFrom:
        secretKeyRef:
          key: SESSION_TIME
          name: 0403724e-d0b5-4edb-ad16-dc310d22b771-env
    - name: USE_POSTGRES_FOR_ACTION_CABLE
      valueFrom:
        secretKeyRef:
          key: USE_POSTGRES_FOR_ACTION_CABLE
          name: 0403724e-d0b5-4edb-ad16-dc310d22b771-env
    - name: WEBSOCKET_PORT
      valueFrom:
        secretKeyRef:
          key: WEBSOCKET_PORT
          name: 0403724e-d0b5-4edb-ad16-dc310d22b771-env
    - name: VCAP_SERVICES
      valueFrom:
        secretKeyRef:
          key: VCAP_SERVICES
          name: 0403724e-d0b5-4edb-ad16-dc310d22b771-vcap-services
    resources: {}
  builder:
    apiVersion: kpack.io/v1alpha2
    kind: ClusterBuilder
    name: cf-kpack-cluster-builder
  cache:
    volume:
      size: 2G
  failedBuildHistoryLimit: 10
  imageTaggingStrategy: BuildNumber
  serviceAccountName: kpack-service-account
  source:
    registry:
      image: europe-west1-docker.pkg.dev/cf-on-k8s-wg/acceptance-images/b039f9ea-3a1f-4c32-9664-4dc01e5238ac@sha256:5928c51ad2fe4a5727aa7fea563af40f049e82d19926919ad58aa43f5e125a33
      imagePullSecrets:
      - name: image-registry-credentials
  successBuildHistoryLimit: 10
  tag: europe-west1-docker.pkg.dev/cf-on-k8s-wg/acceptance-images/kpack/beta/42c5aa03-9d5e-416b-ae92-3aaba8603547
status:
  ...
```

Here is the corresponding pod:

```
apiVersion: v1
kind: Pod
metadata:
  ...
  name: 42c5aa03-9d5e-416b-ae92-3aaba8603547-build-1-build-pod
  namespace: cf-space-4da0295e-53a9-4500-a15e-576aee052999
  ownerReferences:
  - apiVersion: kpack.io/v1alpha2
    blockOwnerDeletion: true
    controller: true
    kind: Build
    name: 42c5aa03-9d5e-416b-ae92-3aaba8603547-build-1
    uid: 3a8116f2-3456-4507-92d6-d81707ad36ed
  uid: e6c5b9b0-210d-4ab4-ad40-7cfae6d36774
spec:
  containers:
    ...
  initContainers:
  ...
  - args:
    - -layers=/layers
    - -app=/workspace
    - -group=/layers/group.toml
    - -plan=/layers/plan.toml
    command:
    - /cnb/lifecycle/builder
    env:
    - name: CNB_PLATFORM_API
      value: "0.8"
    - name: SERVICE_BINDING_ROOT
      value: /platform/bindings
    image: europe-west1-docker.pkg.dev/cf-on-k8s-wg/acceptance-images/kpack/beta@sha256:3caaad0efb8604d621999cd260f09aa9ca31cb81d2fe7864d64f807640de6167
    imagePullPolicy: IfNotPresent
    name: build
    resources: {}
    securityContext:
      allowPrivilegeEscalation: false
      capabilities:
        drop:
        - ALL
      runAsNonRoot: true
      seccompProfile:
        type: RuntimeDefault
    terminationMessagePath: /dev/termination-log
    terminationMessagePolicy: File
    volumeMounts:
    - mountPath: /layers
      name: layers-dir
    - mountPath: /platform
      name: platform-dir
    - mountPath: /workspace
      name: workspace-dir
    - mountPath: /var/run/secrets/kubernetes.io/serviceaccount
      name: kube-api-access-hw5dt
      readOnly: true
    ...
```

As you can see the DATABASE_URL along with many other env vars is not present in the build init container

**Comment by gcapizzi** [comment]
Update: it looks like `builpacks/lifecycle` [filters out](https://github.com/buildpacks/lifecycle/blob/81da25326c90bdf6afcee6cec4ab33afcd3c193c/env/build.go#L26) environment variables that are not in [this list](https://github.com/buildpacks/lifecycle/blob/81da25326c90bdf6afcee6cec4ab33afcd3c193c/env/build.go#L8-L18). This might be the reason the environment is not propagated to the build step: it wouldn't make a difference!

I'm not sure in which way Kpack depends on `buildpack/lifecycle`: is it the only admitted lifecycle implementation?

**Comment by matthewmcnew** [comment]
`secretKeyRef` is not supported yet, in kpack: https://github.com/pivotal/kpack/issues/554. I believe that is likely the source of this issue. Closing this issue to track on https://github.com/pivotal/kpack/issues/554.

**Comment by matthewmcnew** [comment]
> Update 3: I found out how pack does it! It creates an [env layer](https://github.com/buildpacks/pack/blob/30de289455817bb90d760da0b900131443185a5d/internal/builder/builder.go#L353) containing a directory (called platform directory) which gets [picked up by lifecycle](https://github.com/buildpacks/lifecycle/blob/81da25326c90bdf6afcee6cec4ab33afcd3c193c/buildpack/build.go#L201) unless the [buildpack is configured to clear the environment](https://github.com/buildpacks/lifecycle/blob/81da25326c90bdf6afcee6cec4ab33afcd3c193c/buildpack/build.go#L198).

Kpack similar to pack will write env variables to the platform directory: https://github.com/pivotal/kpack/blob/main/cmd/build-init/main.go#L160
